pytorch/pytorch_ddp_1machine.py

```diff
@@ -0,0 +1,216 @@
+
+    # Generate the optimizers.
+    optimizer_name = trial.suggest_categorical("optimizer", ["Adam", "RMSprop", "SGD"])
+    lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True)
+    optimizer = getattr(optim, optimizer_name)(model.parameters(), lr=lr)
+
+    # Get the FashionMNIST dataset.
+    train_loader, valid_loader, train_sampler, valid_sampler = get_mnist()
+
+    accuracy = 0
+    # Training of the model.
+    for epoch in range(EPOCHS):
+        model.train()
+        # Shuffle train dataset.
+        train_sampler.set_epoch(epoch)
+        for data, target in train_loader:
+            data, target = data.view(data.size(0), -1).to(device), target.to(device)
+
+            optimizer.zero_grad()
+            output = model(data)
+            loss = F.nll_loss(output, target)
+            loss.backward()
+            optimizer.step()
+
+        # Validation of the model.
+        model.eval()
+        correct = 0
+        with torch.no_grad():
+            for data, target in valid_loader:
+                data, target = data.view(data.size(0), -1).to(device), target.to(device)
+                output = model(data)
+                # Get the index of the max log-probability.
+                pred = output.argmax(dim=1, keepdim=True)
+                correct += pred.eq(target.view_as(pred)).sum().item()
+
+        correct_tensor = torch.tensor([correct], dtype=torch.int).to(device)
+        dist.all_reduce(correct_tensor)
+        total_correct = correct_tensor.item()
+        accuracy = total_correct / len(valid_loader.dataset)
+
+        trial.report(accuracy, epoch)
+
+        # Handle pruning based on the intermediate value.
+        if trial.should_prune():
+            raise optuna.exceptions.TrialPruned(f"Trial was pruned at epoch {epoch}.")
+
+    return accuracy
+
+
+def setup(rank, world_size):
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+    dist.init_process_group("nccl", rank=rank, world_size=world_size)
+
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+def run_optimize(rank, study, world_size, device_ids):
+    devi = device_ids[rank]
+    print(f"Running basic DDP example on rank {rank} device {devi}.")
+
+    # Set environmental variables required by torch.distributed.
+    setup(rank, world_size)
+
+    if rank == 0:
+        study.optimize(
+            partial(objective, device_id=devi),
+            n_trials=N_TRIALS,
+            timeout=300,
+        )
+    else:
+        for _ in range(N_TRIALS):
+            try:
+                objective(None, devi)
+            except optuna.TrialPruned:
+                pass
+
+    cleanup()
+
+
+if __name__ == "__main__":
+
+    # Download dataset before starting the optimization.
+    datasets.FashionMNIST(DIR, train=True, download=True)
+
+    study = optuna.create_study(
+        storage="sqlite:///example.db",
+        direction="maximize",
+        study_name="pytorch_ddp",
+        load_if_exists=True,
```

Comment:
Do we need sqlite for this example? Personally, the example without database would be simpler and easy to use for users.

Reply:
I think just followed the example of [pytorch_lightning_ddp.py](https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_lightning_ddp.py). I got your point, but I guess the most likely next thing after getting the example code run is probably to find a way to log the tuning history, because in most of the cases, one would want to check the tuning history. I don't mind removing this part, but I also feel it would offer some convenience to users. We can make it optional with an argument, for example.

This probably should be a separate topic, but since you mentioned about database, I wonder if optuna could avoid repeating the same (or very similar) parameters that it tried in the past without saving them to a database. I'm asking this because it seems to me that optuna gets new parameters by referencing to the database (or the tuning history) especially when working with multiple GPUs.

Reply:
## `pytorch_lightning_ddp.py`

I've reviewed this code, so I know a little bit of the background of this example and integration module. Due to technical difficulties, we would only support distributed training with a database. This is why the example code uses SQL. So I don't think we need to follow `pytorch_lightning_ddp.py`.  Another drawback to using SQL for storage is its speed performance. Using SQL is widely documented in Optuna, so I believe that Optuna users can extend the example without database storage to with database storage.


## Repeated same/similar params

As an optuna feature, there isn't, but I think we can implement a similar feature: https://github.com/optuna/optuna/issues/1001#issuecomment-862843041 or https://github.com/optuna/optuna/issues/2021#issuecomment-729308016.

Reply:
>we would only support distributed training with a database

Do you mean that database is only supported in _distributed training_ but not in other cases? If so, how should users save tuning history?

>Another drawback to using SQL for storage is its speed performance. 

What would be a better way to store tuning history then? I thought optuna only supports SQL until the latest version 3.1.0.
I also want to know if there're alternative storage options.

>As an optuna feature, there isn't,

Do you mean that optuna doesn't have a feature to check parameters it tired in the past?
I think I got my answer. When I run my code without specifying `storage`, I see it prints out
`A new study created in memory with name: no-name-eefeae74-8d1d-4613-9b57-b18177f10da3`
So it seems like optuna is storing tuning history in memory instead of SQL, and I guess multiple trials can still refer to this storage in the memory. But I get following errors at the end of tuning if I don't specify `storage`.
```
Study statistics: 
  Number of finished trials:  0
  Number of pruned trials:  0
  Number of complete trials:  0
Best trial:
Traceback (most recent call last):
  File ".../optuna-examples/pytorch/pytorch_ddp_1machine.py", line 219, in <module>
    trial = study.best_trial
  File ".../anaconda3/envs/optuna/lib/python3.9/site-packages/optuna/study/study.py", line 159, in best_trial
    return copy.deepcopy(self._storage.get_best_trial(self._study_id))
  File ".../anaconda3/envs/optuna/lib/python3.9/site-packages/optuna/storages/_in_memory.py", line 262, in get_best_trial
    raise ValueError("No trials are completed yet.")
ValueError: No trials are completed yet.
```
So I guess, since I'm using _distributed training_, I still need SQL storage anyway?

Reply:
> Do you mean that database is only supported in distributed training but not in other cases? If so, how should users save tuning history?

I mean, pytorch lighting with data distributed training works only with database.

> For the distributed data parallel training, the version of PyTorchLightning needs to be higher than or equal to v1.5.0. In addition, [Study](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study) should be instantiated with RDB storage.

Reference: https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.PyTorchLightningPruningCallback.html

---

> What would be a better way to store tuning history then?

I suppose saving `study` would be simple to save optimisation history. https://optuna.readthedocs.io/en/stable/faq.html#how-can-i-save-and-resume-studies

Another way is to use optuna's callback feature, such as [mlflow](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.MLflowCallback.html) or [wandb](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.WeightsAndBiasesCallback.html).

---

> Do you mean that optuna doesn't have a feature to check parameters it tired in the past?

Yes.

---

> So I guess, since I'm using distributed training, I still need SQL storage anyway?

I don't think so because another distributed training code: https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_distributed_simple.py, doesn't use database.

Reply:
Sorry, my last answer might be wrong. When we used `spawn` for ddp training for pytorch-lighting we actually had similar issue. So as you said, swamp-based ddp requires databse.

Reply:
>I suppose saving study would be simple to save optimisation history

Thank you for the information. Good to know that study can be saved as pickle. 

>swamp-based ddp might require database storage

I see, thanks for the background information.